### DISCUSSION TITLE
Lost label names after using trtexec to convert an onnx file to engine file

### DISCUSSION THREAD

**Discussion by aurexav** [ultralytics/ultralytics#14941] [Q&A] [answered]
And I try to search how to set the names manually.

If the model file is onnx or engine, these don't work for me.
I can only set the names manually when the model source is pt file.

```py
model.names = ['foo', 'bar']
model.model.names = {0: 'foo', 1: 'bar'}
```

`result.plot()` shows `class 0`, `class 1`, .. Instead of my labels.

---

And I notice that there is an warning `WARNING ⚠️ Metadata not found for 'model=foo.engine'`.

How to keep the metadata after convert the onnx file to engine file?

**Comment by glenn-jocher** [comment]
@AurevoirXavier it seems like you're encountering an issue with metadata retention when converting an ONNX file to an engine file using `trtexec`. The warning `Metadata not found for 'model=foo.engine'` indicates that the metadata, including class names, is not being preserved during the conversion process.

Unfortunately, TensorRT's `trtexec` tool does not support embedding metadata directly into the engine file. This is why you can't set the class names directly on the engine file as you would with a `.pt` file.

A workaround is to manually set the class names in your inference script after loading the engine file. Although this isn't as seamless as having the metadata embedded, it ensures that your labels are correctly applied during inference.

If you continue to face issues, please ensure you are using the latest versions of the Ultralytics and TensorRT packages, as updates may include fixes or improvements related to your problem.

Feel free to reach out if you have any further questions or need additional assistance.

**Comment by aurexav** [comment, marked as answer]
I decided to use onnx to generate the engine and cache the engine file instead of converting it myself. This approach works well for me.

**Reply by glenn-jocher** [reply]
@AurevoirXavier it's great to hear that using ONNX to generate the engine and caching the engine file is working well for you. If you encounter any further issues or need additional assistance, feel free to reach out. Happy coding!

**Reply by nckrwlmn** [reply]
What does that mean though? Show us how to do it. I’m having the same issue converting onnx to engine. Labels aren’t right afterwards.
